DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky/src/memory_transfers.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef __MEMORY_TRANSFERS_HPP__
+#define __MEMORY_TRANSFERS_HPP__
+
+#include "constexpr_math.hpp"
+#include "tuple.hpp"
+#include "unrolled_loop.hpp"
+
+/*
+  Read matrix_count matrices of type TT from DDR by bursts of num_elem_per_bank
+  elements, and write the matrices to the "MatrixPipe" pipe num_elem_per_bank by
+  num_elem_per_bank elements.
+  Repeat this operations "repetitions" times.
+*/
+template <typename TT,            // Datatype of the elements of the matrix
+          int rows,               // Number of rows of the matrix
+          int columns,            // Number of columns of the matrix
+          int num_elem_per_bank,  // Number of TT elements per DDR burst access
+          typename MatrixPipe     // Output matrix pipe
+          >
+void MatrixReadFromDDRToPipe(
+    TT* matrix_ptr,    // Input matrix pointer
+    int matrix_count,  // Number of matrix to read from DDR
+    int repetitions    // Number of times to write the same matrix to the pipe
+) {
+  // We may perform an incomplete memory read if the number of elements per row
+  // is not a multiple of the DDR burst size
+  constexpr bool kIncompleteBurst = (rows % num_elem_per_bank) != 0;
+  constexpr int kExtraIteration = kIncompleteBurst ? 1 : 0;
+  // Number of DDR burst reads of num_elem_per_bank elements required to read a
+  // full column
+  constexpr int kLoopIterPerColumn = rows / num_elem_per_bank + kExtraIteration;
+  // Number of DDR burst reads of num_elem_per_bank to read all the matrices
+  constexpr int kLoopIter = kLoopIterPerColumn * columns;
+  // Size in bits of the loop iterator over kLoopIter iterations
+  constexpr int kLoopIterBitSize = fpga_tools::BitsForMaxValue<kLoopIter + 1>();
+  // Size of a full matrix
+  constexpr int kMatrixSize = rows * columns;
+
+  sycl::device_ptr<TT> matrix_ptr_device(matrix_ptr);
```

Comment:
Would it be useful to make this agnostic to where the memory lives? Make it the callers job to use a raw pointer, or device_ptr, or host_ptr, or whatever.

Reply:
I can help with this.

Reply:
Let's discuss it when you have time, I'm not sure to understand what you are saying

Reply:
discussed offline; this will be done when we move to a more generic version of memory <-> pipe functions.